api/service/log_service.go

```diff
@@ -15,51 +15,239 @@
+	clusterClient, ok := l.clusterClients[options.Cluster]
 	if !ok {
-		return nil, fmt.Errorf("unable to find cluster %s", options.Cluster)
+		return fmt.Errorf("unable to find cluster %s", options.Cluster)
+	}
+
+	namespace := options.Namespace
+
+	labelSelector := l.getLabelSelector(*options)
+	pods, err := clusterClient.Pods(namespace).List(metav1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return err
 	}
 
-	return logClient.Pods(options.Namespace).GetLogs(options.PodName, options.ToKubernetesLogOption()).Stream()
+	go func() {
+		for {
+			allPodLogs := make([]*PodLog, 0)
+
+			for _, pod := range pods.Items {
+				for _, initContainer := range pod.Spec.InitContainers {
+					options.ContainerName = initContainer.Name
+					stream, err := clusterClient.Pods(namespace).GetLogs(pod.Name, options.ToKubernetesLogOption()).Stream()
+					if err != nil {
+						// Error is handled here by logging it rather than returned because the caller usually does not know how to
+						// handle it. Example of what can trigger ListPodLogs error: while the container is being created/terminated
+						// Kubernetes API server will return error when logs are requested. In such case, it is better to return
+						// empty logs and let the caller retry after the container becomes ready eventually.
+						log.Warnf("Failed to ListPodLogs: %s", err.Error())
+						return
+					}
+
+					scanner := bufio.NewScanner(stream)
+					podLogs := make([]*PodLog, 0)
+					for scanner.Scan() {
+						logLine := scanner.Text()
+
+						// A log line from Kubernetes API server will follow this format:
+						// 2020-07-14T07:48:14.191189249Z {"msg":"log message"}
+						timestampIndex := strings.Index(logLine, " ")
+						if timestampIndex < 0 {
+							// Missing expected RFC3339 timstamp in the log line, skip to next line
+							continue
+						}
+						if (len(logLine) - 1) <= timestampIndex {
+							// Empty log message, skip to next log line
+							continue
+						}
+
+						timestamp, err := time.Parse(time.RFC3339, logLine[:timestampIndex])
+						if err != nil {
+							log.Warnf("log message timestamp is not in RFC3339 format: %s", logLine[:timestampIndex])
+							// Log timestamp value from Kube API server has invalid format, skip to next line
+							continue
+						}
+
+						// We require this check because we send (SinceTime - 1sec) to Kube API Server
+						if options.SinceTime != nil && (timestamp == *options.SinceTime || timestamp.Before(*options.SinceTime)) {
+							continue
+						}
+
+						log := &PodLog{
+							Timestamp:     timestamp,
+							Namespace:     namespace,
+							PodName:       pod.Name,
+							ContainerName: options.ContainerName,
+							TextPayload:   logLine[timestampIndex+1:],
+						}
+
+						podLogs = append(podLogs, log)
+					}
+
+					allPodLogs = append(allPodLogs, podLogs...)
+				}
+
+				for _, container := range pod.Spec.Containers {
```

Comment:
the logic below is the same with the above, can you extract that logic into a function

Reply:
Yup. Done that.